src/java/org/apache/cassandra/streaming/compress/CompressedStreamWriter.java

```diff
@@ -79,12 +91,28 @@ public void write(DataOutputStreamPlus out) throws IOException
                 long bytesTransferred = 0;
                 while (bytesTransferred < length)
                 {
-                    final long bytesTransferredFinal = bytesTransferred;
                     final int toTransfer = (int) Math.min(CHUNK_SIZE, length - bytesTransferred);
                     limiter.acquire(toTransfer);
-                    long lastWrite = out.applyToChannel((wbc) -> fc.transferTo(section.left + bytesTransferredFinal, toTransfer, wbc));
-                    bytesTransferred += lastWrite;
-                    progress += lastWrite;
+
+                    ByteBuf outBuf = channel.alloc().directBuffer(toTransfer, toTransfer);
+                    ByteBuffer outNioBuffer = outBuf.nioBuffer(0, toTransfer);
+                    long lastWrite = fc.read(outNioBuffer, section.left + bytesTransferred);
+                    outBuf.writerIndex((int) lastWrite);
+
+                    if (!waitUntilWritable(channel))
+                        return;
+
+                    channel.writeAndFlush(outBuf).addListener(future -> {
```

Comment:
Maybe you shouldn't flush until you are going to block waiting for writability? That could reduce signaling overhead.